### PR TITLE
test: freeze generated code and simulation output in golden corpora

### PR DESCRIPTION
**Pull request by leon-k-martin**
Groundwork for #69. Before touching the symbolic layer, TVBO needs a way to tell "I refactored the internals" apart from "I changed what TVBO produces". Today it cannot.

## The gap

TVBO freezes the *specification* (the YAML under `tvbo/database/`) and checks that models *run* (`test_run_models.py` executes every model for 10 ms). Nothing pins what sits between them: the source `Dynamics.render_code` emits, and the numbers a spec produces.

A refactor of the symbolic layer can leave every YAML byte-identical and every smoke test green while silently changing a generated right-hand side — term order, a printer dispatch, an inlined function, a `Piecewise` branch. The rewrites #69 proposes are exactly the kind of change that slips through.

## What this adds

Two corpora, failing on disjoint classes of regression. Generated code can change for a purely cosmetic reason and leave the numbers untouched; the numbers can move while every emitted character stays identical (a different noise draw, a reordered reduction, an auxiliary evaluated at a different point in the step). Both are needed.

**`tests/reference_data/codegen/`** — the `jax` and `numpy` source for all 108 curated models, 216 references, compared byte for byte. Backend-independent by design (`julia` and `tvb` need a toolchain and are deliberately absent), so it can gate every pull request.

**`tests/reference_data/numerical/`** — frozen output for six small declarative specs, each covering one axis the others cannot:

| Spec | Axis |
| --- | --- |
| `pendulum_euler` | first-order integrator arithmetic |
| `pendulum_heun` | second-order integrator arithmetic |
| `pendulum_noise_seeded` | stochastic draw + seed threading |
| `conditional_piecewise` | `Piecewise` through codegen into a running solver |
| `network_ring_kuramoto` | coupling sum, weight matrix, per-node state axis |
| `modes_reducedset_hmr` | the `mode_dot` / `mode_sum` array primitives |

Coordinate labels are pinned alongside the values, so a silent axis reordering fails too.

## Why this is trustworthy

Determinism was verified before either corpus was written, not assumed: rendering all curated models under two different `PYTHONHASHSEED` values produces identical output, and every spec is byte-identical across processes, seeded noise included. A diff here is always a real behavioural change.

Both corpora were then proven to *fail* on a real change — a sign flip in a model produced a precise two-line diff, and a 1e-7 relative parameter change was caught at 1.686e-10 max absolute difference.

Numerical comparison uses `rtol=1e-9` rather than bit-for-bit, because floating-point results are not portable across architectures and CI does not run on the machine that produced the references. `TOLERANCES` overrides that per spec, so a chaotic system can state its own without loosening the gate for everything else.

## Design notes

`tests/golden.py` owns the shared harness — reference resolution, re-baselining, orphan pruning, mismatch diagnostics, and case/reference reconciliation — so a new corpus is a storage format plus a comparison rather than another copy of the machinery.

`--regenerate-golden` **cannot report success**. It asserts nothing; it overwrites every reference with whatever the current code produces. Regenerated cases skip rather than pass, and `pytest_sessionfinish` fails the session with a red banner. A re-baseline is a change to what TVBO promises and belongs in its own reviewed commit, never folded into the change that caused it.

Reconciliation is keyed on `(model, format)`, not model name — keyed on the name alone, deleting one format's reference or adding a third format would read as full coverage. The model directory resolves through `tvbo.data.registry.database_dir`, so the corpus and the database the rest of TVBO reads cannot disagree about which models exist. Frozen output lives in `expected/` rather than `output/`, because `.gitignore` carries a bare `output` rule — under that name the references would be silently untracked and this suite would pass on CI while asserting nothing.

Both modules are wired into the CI shards in `ci.yml` and `docker.yml`. A gate no workflow runs is no gate.

## Verification

224 passed, 8 skipped. `--regenerate-golden` exits 4. Orphan detection and pruning both confirmed end to end. Whole suite collects 3378 tests; the core shard passes. Corpus is ~1 MB.

The first CI run on Linux is the real arbiter for the numerical tolerances — if a spec proves margin-sensitive on different hardware it gets a `TOLERANCES` entry rather than a global loosening.

### CODE REVIEW

**Comment by leon-k-martin** [comment]
## Cross-ref: this PR is Phase 0 of the #80 / #81 codegen roadmap

While analysing #80 against `dev` ([plan here](https://github.com/virtual-twin/tvbo/issues/80#issuecomment-5203724028)) it became clear this PR is the blocking prerequisite for that work too, not only for #69. Two things worth folding in.

### 1. The description is stale against the branch

The body says *"the `jax` and `numpy` source for all 108 curated models, **216 references**"* and *"`julia` and `tvb` need a toolchain and are **deliberately absent**"*.

The branch head disagrees — and the branch is right:

```
FORMATS = ("jax", "numpy", "julia", "tvb", "tvboptim", "neuroml")     # :49
```
```
108 jax     108 numpy    108 julia    107 tvboptim    107 neuroml    97 tvb   =  635 refs
```

The docstring already carries the corrected rationale (*"emitting Julia, TVB, tvboptim or NeuroML source needs no Julia, no TVB and no jNeuroML, only the templates in this repository"*), so this is just the body lagging ~3× the actual scope. Worth fixing before review — a reviewer reading "216, jax+numpy only" will mis-scope the diff. Happy to push the edit if you want.

### 2. What it covers, and the gap that matters for #81

`_render` is `model.render_code(format=fmt)` (:155) — **model level**. That is the right scope for #69, which is exactly what this PR was built for.

It does not reach `SimulationExperiment.render_code()`, which is where #81's refactor does its damage:

- `tvbo-jax-sim.py.mako` — the `nx.ancestors` / `topological_sort` scheduling pass (#81 work item 5)
- `tvbo-tvboptim-experiment.py.mako` — 4,715 lines
- all three observation templates — #81's dedup target
- everything sweep / optimisation / exploration / stimulus / noise

`test_tvboptim_byte_identity.py` is numerical equivalence under `jax.disable_jit`, orthogonal by design — it would not notice a reordered emission or a renamed local. `test_codegen_style_contract.py` covers one spec.

**Suggestion: keep that out of this PR.** At +369k/-2280 over 1160 files it is large enough, and the experiment-level corpus needs a decision this PR should not have to carry — there are only **19** curated experiments (8 top-level, 9 neuroml, 2 bifurcation) against 108 models, so a per-experiment cross-product is thin coverage of the *feature* space. The specs worth freezing are the ones already built inline inside `test_codegen_sparse_graph`, `test_correlated_noise_codegen`, `test_stimulus_rendering`, `test_bounded_solver_codegen`, `test_streaming_*`, `test_heterogeneous_adapter`.

That is a follow-up PR reusing `tests/golden.py` verbatim — a new `CASES` list, not new harness — and it is what gates #81 Phase 4. Filed against the roadmap in #80.

### 3. One thing this PR could carry cheaply

`cli/_backends.py:6-9` claims `tests/test_cli_backends_match_ontology.py` keeps `BACKENDS` in sync with `tvb-o-axioms.ttl`. **That file does not exist**, and the table has drifted: the TTL declares tvboptim with `DDEIntegration`, `SDDEIntegration`, `ParameterExploration`, `ConductionSpeedInference` plus `DelayHistoryBuffer` / `InterpolatedDelays` / `StaticDelayBuffer` (`tvb-o-axioms.ttl:191-199`); `BACKENDS["tvboptim"]` has none of them.

It is the same category as this PR — freeze a promise that is currently only asserted in a docstring — and it is ~40 lines. Fits here or stands alone; either way it should land before #81 Phase 1 touches that registry.
